[MRG + 1] added a string for FriedmanMSE (instead impurity) when exporting a do…

fzalkow opened the pull request:
This pull request fixes a minor issue when exporting a dot file via tree.export_graphviz

When having `tree._tree.FriedmanMSE` as criterion for a Decision Tree (that is always the case for the weak learners of `ensemble.GradientBoostingClassifier` e.g.) in each node we found `"impurity = ..."`. Now it should be `"friedman_mse = ..."`.

fzalkow committed:
added a string for FriedmanMSE (instead impurity) when exporting a dot file

fzalkow committed:
missed elif

amueller commented:
LGTM. @arjoly @glouppe ?

arjoly commented on sklearn/tree/export.py:
```diff
@@ -208,7 +208,9 @@ def node_to_str(tree, node_id, criterion):
         # Write impurity
         if impurity:
-            if not isinstance(criterion, six.string_types):
+            if criterion.__class__.__name__ == "FriedmanMSE":
```
Could you use isinstance?

arjoly commented on sklearn/tree/export.py:
```diff
@@ -208,7 +208,9 @@ def node_to_str(tree, node_id, criterion):
         if impurity:
-            if not isinstance(criterion, six.string_types):
+            if criterion.__class__.__name__ == "FriedmanMSE":
+                criterion = "friedman_mse"
```
I would keep the original name (FriedmanMSE).

arjoly commented:
+ 1 with a test and comments are addressed.

fzalkow committed:
now uses isinstance and keeps original name (FriedmanMSE)

fzalkow commented:
Thanks for your suggestions! I have implemented them.

arjoly commented:
The test is still missing.

fzalkow committed:
added a test and reverted string to friedman_mse

fzalkow commented:
I added a test. Is this okay this way? (Having not much experience with unit tests so far.)

I reverted the string to `friedman_mse` for the reason that I have to call a DecisionTreeRegressor with `criterion="friedman_mse"` to init an appropriate tree. For those cases the criterion is called this way anyway. For a GradientBoostingClassifier the criterion is determined in its `_fit_stages` method.

So it seems to be more reasonable to have this one as a kind of public label for this criterion. Would you agree?

fzalkow committed:
corrected test: criterion name has to be only in those nodes where there is "samples"

fzalkow committed:
took if clause into regex

arjoly commented on sklearn/tests/test_tree_export.py:
```diff
@@ -0,0 +1,24 @@
+
+from sklearn.utils.testing import assert_in
+
+def test_friedman_mse_in_graphviz():
```
This test should be located in `sklearn/tree/tests/test_export.py`.

arjoly commented on sklearn/tests/test_tree_export.py:
```diff
@@ -0,0 +1,24 @@
+    import re
+    from sklearn.tree import DecisionTreeRegressor, export_graphviz
+    from sklearn.ensemble import GradientBoostingClassifier
+    from sklearn.externals.six import StringIO
```
Those import should be at the top.

arjoly commented on sklearn/tests/test_tree_export.py:
```diff
@@ -0,0 +1,24 @@
+        export_graphviz(estimator[0], out_file=dot_data)
+
+    for finding in re.finditer("\[.*?samples.*?\]", dot_data.getvalue()):
+        assert_in("friedman_mse", finding.group())
```
This could be written simpler by checking exact match.
NB: There is an assert helper for regex.

arjoly commented on sklearn/tests/test_tree_export.py:
```diff
@@ -0,0 +1,24 @@
+    dot_data = StringIO()
+    export_graphviz(clf, out_file=dot_data)
+
+    clf = GradientBoostingClassifier(n_estimators=2)
```
Can you set the seed for reproducbility?

arjoly commented on sklearn/tests/test_tree_export.py:
```diff
@@ -0,0 +1,24 @@
+    from sklearn.ensemble import GradientBoostingClassifier
+    from sklearn.externals.six import StringIO
+
+    clf = DecisionTreeRegressor(criterion="friedman_mse")
```
Can you set the seed for reproducibility?

arjoly commented:
Looks good whenever the last tid bits are addressed.

I won't have access to internet for the next week. So don't wait for me to merge.

fzalkow committed:
implemented improvement suggestions

fzalkow commented:
Thx for improvement suggestions and have a happy internet-free time.

larsmans commented on sklearn/tree/tests/test_export.py:
```diff
@@ -223,3 +224,56 @@ def test_graphviz_errors():
     # Check class_names error
     out = StringIO()
     assert_raises(IndexError, export_graphviz, clf, out, class_names=[])
+
```
Two blank lines between functions (pep8).

larsmans commented:
Test failure on Windows: the tests wants `value = 0.0`, but gets `value = -0.0`.

amueller commented:
whoops it seems I'm going blind :-/

fzalkow committed:
added newline (for pep8) and reverted to regex solution due to 0.0/-0.0 problem on windows

fzalkow committed:
fixes

fzalkow commented:
For now I just reverted to my old regex solution instead of comparing the whole string. As far as I see for fixing this 0.0/-0.0 inequality, one would have to add 0.0 to all numbers that are formatted when creating the dot string...

	>>> x = -0.0
	>>> print x
	-0.0
	>>> x += 0.0
	>>> print x
	0.0

ogrisel commented:
LGTM, merging. thanks for the fix @fzalkow!
